File: backend/app/models/prospect.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base


def _now() -> datetime:
    return datetime.now(timezone.utc)


class Prospect(Base):
    __tablename__ = "prospects"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    campaign_id: Mapped[str] = mapped_column(String(36), ForeignKey("campaigns.id", ondelete="CASCADE"))

    # Identity — from Apollo
    first_name: Mapped[str] = mapped_column(String(100))
    last_name: Mapped[str] = mapped_column(String(100))
    email: Mapped[str] = mapped_column(String(255))
    title: Mapped[str] = mapped_column(String(150))
    company: Mapped[str] = mapped_column(String(150))
    company_size: Mapped[str | None] = mapped_column(String(30), nullable=True)
    country: Mapped[str] = mapped_column(String(100))
    linkedin_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    # Enrichment — from Tavily + LLM
    recent_news: Mapped[str | None] = mapped_column(Text, nullable=True)
    job_postings: Mapped[list] = mapped_column(JSON, default=list)
    pain_points: Mapped[list] = mapped_column(JSON, default=list)
    personalisation_hook: Mapped[str | None] = mapped_column(Text, nullable=True)
    enriched_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now)

    campaign: Mapped["Campaign"] = relationship(back_populates="prospects")  # noqa: F821
    emails: Mapped[list["Email"]] = relationship(back_populates="prospect", cascade="all, delete-orphan")  # noqa: F821

File: specs/data-models/prospect.ts
// Prospect data models
// A Prospect is one person discovered and enriched by the Researcher Agent.
// Prospects are always scoped to a Campaign (one prospect can't appear in two campaigns).

// ── Core Model ─────────────────────────────────────────────────────────────

export interface Prospect {
  id: string;                        // UUID
  campaign_id: string;               // UUID — parent campaign

  // ── Identity (from Apollo) ───────────────────────────────────────────────
  first_name: string;
  last_name: string;
  email: string;
  title: string;                     // e.g. "VP of Sales"
  company: string;
  company_size: string | null;       // e.g. "51-200", "201-500"
  country: string;                   // e.g. "UAE", "Saudi Arabia"
  linkedin_url: string | null;

  // ── Enrichment (from Tavily + LLM synthesis) ─────────────────────────────
  recent_news: string | null;        // one-sentence company news summary
  job_postings: string[];            // open roles at the company (max 3)
  pain_points: string[];             // inferred by Researcher agent (2–3 items)
  personalisation_hook: string | null; // opening line the Writer uses in Email 1

  enriched_at: string;               // ISO 8601 — when Tavily research completed
}

// ── Derived Helpers ────────────────────────────────────────────────────────

/** Full display name */
export const getProspectFullName = (p: Prospect): string =>
  `${p.first_name} ${p.last_name}`;

/** Initials for avatar fallback */
export const getProspectInitials = (p: Prospect): string =>
  `${p.first_name[0]}${p.last_name[0]}`.toUpperCase();

/** True if the prospect has enough enrichment data to personalise well */
export const isWellEnriched = (p: Prospect): boolean =>
  p.personalisation_hook !== null &&
  p.pain_points.length >= 1 &&
  p.recent_news !== null;

// ── List Response ──────────────────────────────────────────────────────────

export interface ProspectListResponse {
  items: Prospect[];
  total: number;
  page: number;
  limit: number;
}

// ── Company Size Options (for display / filtering) ─────────────────────────

export const COMPANY_SIZE_ORDER = [
  "1-10",
  "11-50",
  "51-200",
  "201-500",
  "501-1000",
  "1001-5000",
  "5001+",
] as const;

export type CompanySize = (typeof COMPANY_SIZE_ORDER)[number];
